compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -519,17 +529,17 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
             let b_offset = a.value.size(self).align_to(b.value.align(self).abi);
             let pair_ty = place.layout.llvm_type(self);
 
-            let mut load = |i, scalar: abi::Scalar, align| {
+            let mut load = |i, scalar: abi::Scalar, layout, align, offset| {
                 let llptr = self.struct_gep(pair_ty, place.llval, i as u64);
                 let llty = place.layout.scalar_pair_element_llvm_type(self, i, false);
                 let load = self.load(llty, llptr, align);
-                scalar_load_metadata(self, load, scalar);
+                scalar_load_metadata(self, load, scalar, layout, offset);
                 self.to_immediate_scalar(load, scalar)
             };
 
             OperandValue::Pair(
-                load(0, a, place.align),
-                load(1, b, place.align.restrict_for_offset(b_offset)),
+                load(0, a, place.layout, place.align, Size::ZERO),
+                load(1, b, place.layout, place.align.restrict_for_offset(b_offset), b_offset),
```

Comment:
This doesn't look right to me. Why would the offset of the second pair element be relevant for the alignment of the value it points to?

Reply:
Both elements of a scalar pair share the same `TyAndLayout`, the offset is used by `layout.pointee_info_at` to determine which element of the pair is being loaded.

This is ~the same logic as is used to generate `align` attributes for scalar pair arguments--
`adjust_for_rust_scalar` calls `pointee_info_at` here: https://github.com/rust-lang/rust/blob/9fbff89354ba7a4ec2d1191a3f405cbe408370f6/compiler/rustc_middle/src/ty/layout.rs#L3070
...which is called from the closure passed to `ArgAbi::new`: https://github.com/rust-lang/rust/blob/9fbff89354ba7a4ec2d1191a3f405cbe408370f6/compiler/rustc_middle/src/ty/layout.rs#L3135
...which is in turn called for both parts of a scalar pair (with the same layout and different offsets) here: https://github.com/rust-lang/rust/blob/9556d7a09a7e80e84430071b29d118f39dde8205/compiler/rustc_target/src/abi/call/mod.rs#L458-L461